File: 48-rotate-image/rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        # Idea: Flip the rows in matrix, then transpose matrix!
        N = len(matrix)

        # Step 1: Transpose the matrix (can only modify matrix)
        for i in range(N):
            for j in range(i + 1, N):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        
        # Step 2: Flip the rows in matrix (can only modify matrix)
        for i in range(N):
            l, r = 0, N - 1
            while l < r:
                matrix[i][l], matrix[i][r] = matrix[i][r], matrix[i][l]
                l += 1
                r -= 1

        return

File: 48-rotate-image/rotate-image.cs
public class Solution {
    public void Rotate(int[][] matrix) {
        // Idea: Transpose, then reverse rows

        // First, transpose matrix
        int n = matrix.Length;
        for (int i = 0; i < n; i++)
            for (int j = i + 1; j < n; j++)
            {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = tmp;
            }
        
        // Now, reverse rows
        for (int i = 0; i < n; i++)
        {
            int[] row = matrix[i];
            int left = 0, right = n - 1;
            while (left < right)
            {
                // Swap logic
                int tmp = row[left];
                row[left] = row[right];
                row[right] = tmp;

                // Loop invariant
                left++; right--;
            }
        }

        return;
    }
}